File: infitu/nodes/continuous_parameter_interface.cpp
#include <ros/ros.h>
#include "infitu/ContinuousParamPolicy.h"
#include "argus_utils/utils/ParamUtils.h"
#include "percepto_msgs/SetParameters.h"

using namespace argus;
using namespace percepto;

class ContinuousParameterInterfaceNode
{
public:

	ContinuousParameterInterfaceNode( ros::NodeHandle& nh,
	                                  ros::NodeHandle& ph )
	{
		_interface.Initialize( nh, ph );

		std::vector<double> init;
		if( GetParam( ph, "initial_parameters", init ) &&
		    !SetOutput( GetVectorView( init ) ) )
		{
			throw std::runtime_error( "Could not set initial parameters." );
		}

		_setterServer = ph.advertiseService( "set_parameters", 
		                                     &ContinuousParameterInterfaceNode::SetterCallback,
		                                     this );
	}

private:

	ContinuousParamPolicy _interface;
	ros::ServiceServer _setterServer;

	bool SetterCallback( percepto_msgs::SetParameters::Request& req,
	                     percepto_msgs::SetParameters::Response& res )
	{
		return SetOutput( GetVectorView( req.parameters ) );
	}

	bool SetOutput( const VectorType& out )
	{
		if( out.size() != _interface.GetNumOutputs() )
		{
			ROS_ERROR_STREAM( "Received " << out.size() << " parameters but interface takes " <<
			                  _interface.GetNumOutputs() );
			return false;
		}
		_interface.SetOutput( out );
		return true;
	}
};

int main( int argc, char** argv )
{
	ros::init( argc, argv, "continuous_parameter_interface" );

	ros::NodeHandle nh, ph( "~" );
	ContinuousParameterInterfaceNode cppn( nh, ph );
	ros::spin();
	return 0;
}

File: infitu/nodes/ContinuousParameterInterface.py
#!/usr/bin/env python

import rospy
import infitu

from percepto_msgs.srv import SetParameters
from percepto_msgs.srv import GetParameters, GetParametersResponse


class ContinuousParameterInterface(object):
    def __init__(self):
        interface_info = rospy.get_param('~interface')
        self.interface = infitu.parse_interface(interface_info)

        self.set_param = rospy.Service('~set_parameters',
                                       SetParameters,
                                       self.set_param_callback)
        self.get_param = rospy.Service('~get_parameters',
                                       GetParameters,
                                       self.get_param_callback)

    def set_param_callback(self, req):
        try:
            self.interface.set_values(req.parameters)
            return []
        except ValueError as e:
            rospy.logerr('Could not set params: %s', str(e))
            return None

    def get_param_callback(self, req):
        req = GetParametersResponse()
        req.parameters = self.interface.get_values()
        return req


if __name__ == '__main__':
    rospy.init_node('continuous_parameter_interface')
    cpi = ContinuousParameterInterface()
    try:
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
